File: app/src/main/java/com/jakebarnby/filemanager/activities/source/SourceFragment.java
package com.jakebarnby.filemanager.activities.source;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.provider.Settings;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AnimationUtils;
import android.widget.Button;
import android.widget.HorizontalScrollView;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.util.ViewPreloadSizeProvider;
import com.jakebarnby.filemanager.R;
import com.jakebarnby.filemanager.activities.source.adapters.FileSystemAdapter;
import com.jakebarnby.filemanager.activities.source.adapters.FileSystemGridAdapter;
import com.jakebarnby.filemanager.activities.source.adapters.FileSystemListAdapter;
import com.jakebarnby.filemanager.managers.SelectedFilesManager;
import com.jakebarnby.filemanager.models.Source;
import com.jakebarnby.filemanager.models.files.SourceFile;
import com.jakebarnby.filemanager.services.SourceTransferService;
import com.jakebarnby.filemanager.util.Constants;
import com.jakebarnby.filemanager.util.TreeNode;
import com.jakebarnby.filemanager.util.Utils;

/**
 * A placeholder fragment containing a simple view.
 */
public abstract class SourceFragment extends Fragment {

    private Source                  mSource;

    protected RecyclerView          mRecycler;
    protected FileSystemListAdapter mFileSystemListAdapter;
    protected FileSystemGridAdapter mFileSystemGridAdapter;
    protected ProgressBar           mProgressBar;
    protected Button                mConnectButton;
    protected ImageView             mSourceLogo;
    protected View                  mDivider;

    private LinearLayout            mBreadcrumbBar;
    private HorizontalScrollView    mBreadcrumbWrapper;

    /**
     * Authenticate the current mSource
     */
    protected abstract void authenticateSource();

    /**
     * Load the current mSource
     */
    protected abstract void loadSource();

    public Source getSource() {
        return mSource;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_source, container, false);
        mBreadcrumbWrapper = rootView.findViewById(R.id.breadcrumb_wrapper);
        mBreadcrumbBar = rootView.findViewById(R.id.breadcrumbs);
        mRecycler = rootView.findViewById(R.id.recycler_local);
        mProgressBar = rootView.findViewById(R.id.animation_view);
        mDivider = rootView.findViewById(R.id.divider);
        mConnectButton = rootView.findViewById(R.id.btn_connect);
        mSourceLogo = rootView.findViewById(R.id.image_source_logo);
        mSource = new Source(getArguments().getString(Constants.FRAGMENT_TITLE));

        if (hasToken(mSource.getSourceName())) {
            mConnectButton.setVisibility(View.GONE);
            mSourceLogo.setVisibility(View.GONE);
        }
        mConnectButton.setOnClickListener(v -> {
            if (Utils.isConnectionReady(SourceFragment.this.getContext())) {
                authenticateSource();
            } else {
                Snackbar.make(rootView, R.string.err_no_connection, Snackbar.LENGTH_LONG).show();
            }
        });
        return rootView;
    }

    public void setMultiSelectEnabled(boolean enabled) {
        if (enabled) {
            ((SourceActivity)getActivity()).toggleFloatingMenu(true);
        }
        this.mSource.setMultiSelectEnabled(enabled);
        if (mRecycler.getAdapter() != null) {
            ((FileSystemAdapter) mRecycler.getAdapter()).setMultiSelectEnabled(enabled);
            mRecycler.getAdapter().notifyDataSetChanged();
        }
    }

    /**
     * Reload the {@link RecyclerView}
     */
    public void refreshRecycler() {
        mRecycler.getAdapter().notifyDataSetChanged();
    }

    /**
     * Checks if this mSource has a valid access token
     * @return  Whether there is a valid access token for this mSource
     */
    protected boolean hasToken(String sourceName) {
        SharedPreferences prefs = getActivity().getPreferences(Context.MODE_PRIVATE);
        String accessToken = prefs.getString(sourceName + "-access-token", null);
        return accessToken != null;
    }

    /**
     * Set the {@link RecyclerView} layout and adapter based on users preferences
     */
    public void initRecyclerView() {
        FileSystemAdapter newAdapter;

        SharedPreferences sharedPref = getActivity().getPreferences(Context.MODE_PRIVATE);
        String viewType = sharedPref.getString("ViewAs", "List");
        if (viewType.equals("List")) {
            mRecycler.setLayoutManager(new LinearLayoutManager(getContext(),
                    LinearLayoutManager.VERTICAL,
                    false));
            newAdapter = mFileSystemListAdapter;
        } else {
            mRecycler.setLayoutManager(new GridLayoutManager(getContext(), 4));
            newAdapter = mFileSystemGridAdapter;
        }

        mRecycler.setAdapter(newAdapter);
        if (mRecycler.getAdapter() != null) {
            mRecycler.getAdapter().notifyDataSetChanged();
        }
        mDivider.setVisibility(View.VISIBLE);
    }

    /**
     * Initialize the UI for this mSource
     * @param file                  The root file of the mSource
     * @param onClickListener       The click listener for files and folders
     * @param onLongClickListener   The long click listener for files and folders
     */
    protected void initAdapters(TreeNode<SourceFile> file,
                                FileSystemAdapter.OnFileClickedListener onClickListener,
                                FileSystemAdapter.OnFileLongClickedListener onLongClickListener) {
        mFileSystemListAdapter = new FileSystemListAdapter(file,
                                                           Glide.with(this).asDrawable(),
                                                           new ViewPreloadSizeProvider());
        mFileSystemListAdapter.setOnClickListener(onClickListener);
        mFileSystemListAdapter.setOnLongClickListener(onLongClickListener);

        mFileSystemGridAdapter = new FileSystemGridAdapter(file,
                                                           Glide.with(this).asDrawable(),
                                                           new ViewPreloadSizeProvider());
        mFileSystemGridAdapter.setOnClickListener(onClickListener);
        mFileSystemGridAdapter.setOnLongClickListener(onLongClickListener);

        initRecyclerView();
    }

    /**
     * Opens the app details page for this app
     */
    protected void showAppDetails() {
        Intent intent = new Intent();
        intent.setAction(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
        Uri uri = Uri.fromParts("package", "com.jakebarnby.filemanager", null);
        intent.setData(uri);
        startActivity(intent);
    }

    /**
     * Construct a click listener for a file or folder
     * @return  The constructed listener
     */
    protected FileSystemAdapter.OnFileClickedListener createOnClickListener() {
        return (file, isChecked, position) -> {
            if (mSource.isMultiSelectEnabled()) {
                if (isChecked) {
                    SelectedFilesManager
                            .getInstance()
                            .getSelectedFiles(SelectedFilesManager.getInstance().getOperationCount())
                            .add(file);
                } else {
                    SelectedFilesManager
                            .getInstance()
                            .getSelectedFiles(SelectedFilesManager.getInstance().getOperationCount())
                            .remove(file);
                }

                int size = SelectedFilesManager
                        .getInstance()
                        .getSelectedFiles(SelectedFilesManager.getInstance().getOperationCount())
                        .size();
                getActivity().setTitle(size + " selected");
                //TODO: Set the Fragment tab title with selected count, e.g. LOCAL (3) DROPBOX (1)
            } else {
                if (file.getData().isDirectory()) {
                    ((FileSystemAdapter) mRecycler.getAdapter()).setCurrentDirectory(file);
                    mSource.setCurrentDirectory(file);
                    ((SourceActivity)getActivity()).getSourceManager().setActiveDirectory(file);
                    pushBreadcrumb(file);
                    mRecycler.getAdapter().notifyDataSetChanged();
                } else {
//                    if (!((SourceActivity) getActivity()).startFreeSpaceCheck() &&
//                            !getSourceName().equals(Constants.Sources.LOCAL)) {
//                        return;
//                    }

                    if (Utils.getFreeSpace(Environment.getExternalStorageDirectory()) > file.getData().getSize()) {
                        SelectedFilesManager.getInstance().addNewSelection();
                        ((SourceActivity)getActivity()).getSourceManager().addFileAction(
                                SelectedFilesManager.getInstance().getOperationCount()-1,
                                SourceActivity.FileAction.OPEN);
                        SourceTransferService.startActionOpen(getContext(), file.getData());
                    } else {
                        Snackbar.make(mRecycler,
                                String.format(getString(R.string.err_no_free_space), file.getData().getSourceName()),
                                Snackbar.LENGTH_LONG).show();
                    }
                }
            }
        };
    }

    /**
     * Construct a long click listener for a file or folder
     * @return  The constructed listener
     */
    protected FileSystemAdapter.OnFileLongClickedListener createOnLongClickListener() {
        return file -> {
            if (!mSource.isMultiSelectEnabled()) {
                setMultiSelectEnabled(true);

                if (SelectedFilesManager.getInstance().getOperationCount() == 0) {
                    SelectedFilesManager.getInstance().addNewSelection();
                }

                SelectedFilesManager
                        .getInstance()
                        .getSelectedFiles(SelectedFilesManager.getInstance().getOperationCount())
                        .add(file);

                int size = SelectedFilesManager
                        .getInstance()
                        .getSelectedFiles(SelectedFilesManager.getInstance().getOperationCount())
                        .size();
                getActivity().setTitle(size + " selected");
            }
        };
    }

    /**
     * Push a breadcrumb onto the view stack
     * @param directory The directory to push
     */
    protected void pushBreadcrumb(TreeNode<SourceFile> directory) {
        final ViewGroup crumbLayout = (ViewGroup) getActivity().getLayoutInflater()
                .inflate(R.layout.view_breadcrumb, null);

        crumbLayout.findViewById(R.id.crumb_arrow)
                .setVisibility(directory.getParent() == null ? View.GONE : View.VISIBLE);

        TextView text = crumbLayout.findViewById(R.id.crumb_text);
        text.setText(directory.getParent() == null ?
                directory.getData().getSourceName().substring(0,1).toUpperCase() + directory.getData().getSourceName().substring(1).toLowerCase():
                directory.getData().getName());

        crumbLayout.setOnClickListener(v -> {
            TextView crumbText =  v.findViewById(R.id.crumb_text);

            int diff = (mBreadcrumbBar.getChildCount()-1) - mBreadcrumbBar.indexOfChild(v);
            for (int i = 0; i < diff; i++) popBreadcrumb();

            String name = crumbText.getText().toString();
            if (mSource.getCurrentDirectory().getData().getName().equals(name)) return;
            TreeNode<SourceFile> selectedParent = TreeNode.findParent(mSource.getCurrentDirectory(), name);

            ((SourceActivity)getActivity()).getSourceManager().setActiveDirectory(selectedParent);
            ((FileSystemAdapter)mRecycler.getAdapter()).setCurrentDirectory(selectedParent);
            mSource.setCurrentDirectory(selectedParent);
            mRecycler.getAdapter().notifyDataSetChanged();
        });
        mBreadcrumbWrapper.postDelayed(() ->
                mBreadcrumbWrapper.fullScroll(HorizontalScrollView.FOCUS_RIGHT),
                50L);
        mBreadcrumbBar.addView(crumbLayout);
        crumbLayout.startAnimation(AnimationUtils.loadAnimation(getContext(), R.anim.breadcrumb_overshoot_z));
    }

    /**
     * Pop the latest breakcrumb added to the stack
     */
    protected void popBreadcrumb() {
        mBreadcrumbBar.removeViewAt(mBreadcrumbBar.getChildCount()-1);
    }

    /**
     * If performing an action on a non-local directory, check internet and
     */
    protected boolean checkConnectionStatus() {
        if (!mSource.getSourceName().equals(Constants.Sources.LOCAL)) {
            if (!Utils.isConnectionReady(getContext())) {
                Snackbar.make(mRecycler, R.string.err_no_connection, Snackbar.LENGTH_LONG).show();
                return false;
            }
        }
        return true;
    }
}

File: app/src/main/java/com/jakebarnby/filemanager/ui/sources/SourceFragment.kt
package com.jakebarnby.filemanager.ui.sources

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.provider.Settings
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.widget.*
import androidx.appcompat.app.AlertDialog
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import com.jakebarnby.filemanager.R
import com.jakebarnby.filemanager.models.SourceFile
import com.jakebarnby.filemanager.models.SourceType
import com.jakebarnby.filemanager.models.ViewType
import com.jakebarnby.filemanager.services.SourceTransferService
import com.jakebarnby.filemanager.sources.googledrive.GoogleDriveFragment
import com.jakebarnby.filemanager.sources.googledrive.GoogleDriveSource
import com.jakebarnby.filemanager.sources.local.LocalFragment
import com.jakebarnby.filemanager.sources.onedrive.OneDriveFragment
import com.jakebarnby.filemanager.sources.onedrive.OneDriveSource
import com.jakebarnby.filemanager.ui.adapters.FileAdapter
import com.jakebarnby.filemanager.ui.adapters.FileAdapter.OnFileLongClickedListener
import com.jakebarnby.filemanager.ui.adapters.FileDetailedListAdapter
import com.jakebarnby.filemanager.ui.adapters.FileGridAdapter
import com.jakebarnby.filemanager.ui.adapters.FileListAdapter
import com.jakebarnby.filemanager.util.*
import com.jakebarnby.filemanager.util.Constants.GRID_SIZE
import com.jakebarnby.filemanager.util.Constants.Prefs
import dagger.android.support.DaggerFragment
import java.util.*
import javax.inject.Inject

abstract class SourceFragment : DaggerFragment(), SourceFragmentContract.View {

    @Inject
    protected lateinit var presenter: SourceFragmentContract.Presenter

    var recycler: RecyclerView? = null

    private val adapters = mutableMapOf<ViewType, FileAdapter>()

    private lateinit var progressBar: ProgressBar
    private lateinit var connectButton: Button
    private lateinit var sourceLogo: ImageView
    private lateinit var divider: View

    private lateinit var breadcrumbBar: LinearLayout
    private lateinit var breadcrumbWrapper: HorizontalScrollView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        presenter.subscribe(this)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        presenter.checkState()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.fragment_source, container, false)

        breadcrumbWrapper = rootView.findViewById(R.id.breadcrumb_wrapper)
        breadcrumbBar = rootView.findViewById(R.id.breadcrumbs)
        recycler = rootView.findViewById(R.id.recycler_local)
        progressBar = rootView.findViewById(R.id.animation_view)
        divider = rootView.findViewById(R.id.divider_sort)
        connectButton = rootView.findViewById(R.id.btn_connect)
        sourceLogo = rootView.findViewById(R.id.img_source_logo)

        connectButton.setOnClickListener {
            presenter.onConnect()
        }

        sourceLogo.setImageResource(
            Utils.resolveLogoId(presenter.source.sourceId)
        )
        return rootView
    }

    override fun onResume() {
        super.onResume()
        presenter.subscribe(this)
    }

    override fun onPause() {
        presenter.unsubscribe()
        super.onPause()
    }

    override fun authenticate() {
        if (this is OneDriveFragment) {
            (presenter.source as? OneDriveSource)?.authenticateSource(this)
        } else {
            presenter.source.authenticate(context!!)
        }
    }

    override fun startActionOpen(toOpen: SourceFile) {
        SourceTransferService.startActionOpen(context!!, toOpen)
    }

    override fun setSelectedCountTitle(size: Int) {
        activity?.title = String.format(
            Locale.getDefault(),
            getString(R.string.selected_title),
            size
        )
    }

    override fun updateFileList() {
        recycler?.adapter?.notifyDataSetChanged()
    }

    override fun showLoadError(sourceName: Int) {
        val message = String.format(
            "%s %s %s",
            getString(R.string.problem_encountered),
            getString(R.string.loading_source),
            "$sourceName files."
        )
        AlertDialog.Builder(context!!)
            .setTitle(R.string.dialog_error)
            .setMessage(message)
            .setNegativeButton(R.string.close) { dialog, _ -> dialog.dismiss() }
            .create()
            .show()
    }

    override fun showProgressBar() {
        progressBar.visibility = View.VISIBLE
    }

    override fun hideProgressBar() {
        progressBar.visibility = View.GONE
    }

    override fun showConnectButton() {
        connectButton.visibility = View.VISIBLE
    }

    override fun hideConnectButton() {
        connectButton.visibility = View.GONE
    }

    override fun showSourceLogo() {
        sourceLogo.visibility = View.VISIBLE
    }

    override fun hideSourceLogo() {
        sourceLogo.visibility = View.GONE
    }

    override fun showFileList() {
        TODO("Not yet implemented")
    }

    override fun hideFileList() {
        recycler?.visibility = View.GONE
        divider.visibility = View.GONE
        breadcrumbWrapper.visibility = View.GONE
        sourceLogo.visibility = View.VISIBLE
        connectButton.visibility = View.VISIBLE
        progressBar.visibility = View.GONE
    }

    override fun showNoConnectionSnackBar() {
        Snackbar.make(recycler!!, R.string.err_no_connection, Snackbar.LENGTH_LONG).show()
    }

    override fun showNotEnoughSpaceSnackBar() {
        Snackbar.make(recycler!!, R.string.err_no_free_space, Snackbar.LENGTH_LONG).show()
    }

    fun initRecyclerView() {
        val viewType = ViewType.getFromValue(presenter.prefsManager.getInt(
            Prefs.VIEW_TYPE_KEY,
            ViewType.LIST.value
        ))

        TreeNode.sortTree(
            presenter.source.rootNode,
            Comparators.resolveComparatorForPrefs(presenter.prefsManager)
        )

        val newAdapter = when (viewType) {
            ViewType.LIST -> {
                recycler?.layoutManager = LinearLayoutManager(
                    context,
                    LinearLayoutManager.VERTICAL,
                    false
                )
                adapters[ViewType.LIST]
            }
            ViewType.DETAILED_LIST -> {
                recycler?.layoutManager = LinearLayoutManager(
                    context,
                    LinearLayoutManager.VERTICAL,
                    false
                )
                adapters[ViewType.DETAILED_LIST]
            }
            ViewType.GRID -> {
                recycler?.layoutManager = GridLayoutManager(
                    context,
                    GRID_SIZE
                )
                adapters[ViewType.GRID]
            }
            else -> {
                recycler?.layoutManager = LinearLayoutManager(
                    context,
                    LinearLayoutManager.VERTICAL,
                    false
                )
                adapters[ViewType.LIST]
            }
        }
        recycler?.adapter = newAdapter
        recycler?.adapter?.notifyDataSetChanged()
        recycler?.visibility = View.VISIBLE
        divider.visibility = View.VISIBLE
        breadcrumbWrapper.visibility = View.VISIBLE
    }

    override fun populateList() {
        val clickListener = object : FileAdapter.OnFileClickedListener {
            override fun onClick(file: TreeNode<SourceFile>, isChecked: Boolean, position: Int) {
                presenter.onFileSelected(file, isChecked, position)
            }
        }
        val longClickListener = object : OnFileLongClickedListener {
            override fun onLongClick(file: TreeNode<SourceFile>) {
                presenter.onFileLongSelected(file)
            }
        }
        val listAdapter = FileListAdapter(
            presenter.source,
            presenter.selectedFilesManager,
            presenter.prefsManager
        ).apply {
            setOnClickListener(clickListener)
            setOnLongClickListener(longClickListener)
        }
        val gridAdapter = FileGridAdapter(
            presenter.source,
            presenter.selectedFilesManager,
            presenter.prefsManager
        ).apply {
            setOnClickListener(clickListener)
            setOnLongClickListener(longClickListener)
        }
        val detailedListAdapter = FileDetailedListAdapter(
            presenter.source,
            presenter.selectedFilesManager,
            presenter.prefsManager
        ).apply {
            setOnClickListener(clickListener)
            setOnLongClickListener(longClickListener)
        }

        adapters[ViewType.LIST] = listAdapter
        adapters[ViewType.DETAILED_LIST] = detailedListAdapter
        adapters[ViewType.GRID] = gridAdapter

        initRecyclerView()
    }

    /**
     * Push a breadcrumb onto the view stack
     * @param directory The directory to push
     */
    override fun pushBreadCrumb(
        directory: TreeNode<SourceFile>,
        arrowVisible: Boolean,
        name: String
    ) {
        val crumbLayout = activity
            ?.layoutInflater
            ?.inflate(R.layout.view_breadcrumb, breadcrumbWrapper, false)
            as? ViewGroup
            ?: return

        crumbLayout.findViewById<View>(R.id.crumb_arrow).visibility = if (arrowVisible) {
            View.VISIBLE
        } else {
            View.GONE
        }

        crumbLayout.findViewById<TextView>(R.id.crumb_text)
            ?.text = name

        crumbLayout.setOnClickListener { v: View ->
            val crumbsToPop = breadcrumbBar.childCount - 1 - breadcrumbBar.indexOfChild(v)
            val selectedDirName = v.findViewById<TextView>(R.id.crumb_text).text.toString()

            presenter.onBreadCrumbSelected(selectedDirName, crumbsToPop)
        }

        breadcrumbWrapper.afterLayout {
            breadcrumbWrapper.fullScroll(HorizontalScrollView.FOCUS_RIGHT)
        }
        breadcrumbBar.addView(crumbLayout)

        crumbLayout.startAnimation(
            AnimationUtils.loadAnimation(context, R.anim.breadcrumb_overshoot_z)
        )
    }

    fun pushAllBreadCrumbs(directory: TreeNode<SourceFile>) {
        val breadCrumbs = Stack<TreeNode<SourceFile>>()
        var root = directory
        breadCrumbs.push(root)
        while (root.parent != null) {
            root = root.parent!!
            breadCrumbs.push(root)
        }
        while (breadCrumbs.size > 0) {
            val node = breadCrumbs.pop()
            pushBreadCrumb(
                node,
                node.parent != null,
                if (node.parent == null) {
                    SourceType.values()[node.data.sourceId].sourceName
                } else {
                    node.data.name
                }
            )
        }
    }

    override fun popBreadCrumb() {
        breadcrumbBar.removeViewAt(breadcrumbBar.childCount - 1)
    }

    override fun popAllBreadCrumbs() {
        breadcrumbBar.removeAllViews()
    }

    override fun onCheckPermissions(name: String, requestCode: Int) {
        val activity = activity as? SourceActivity ?: return

        val permissionCheck = ContextCompat.checkSelfPermission(context!!, name)
        if (permissionCheck != PackageManager.PERMISSION_GRANTED) {
            if (!activity.isCheckingPermissions) {
                activity.isCheckingPermissions = true
                requestPermissions(arrayOf(name), requestCode)
            }
        } else {
            if (this is LocalFragment) {
                presenter.source.isLoggedIn = true
                presenter.source.loadFiles(activity)
            } else if (this is GoogleDriveFragment) {
                (presenter.source as GoogleDriveSource)
                    .authGoogle(this)
            }
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        when (requestCode) {
            Constants.RequestCodes.STORAGE_PERMISSIONS -> {
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    presenter.source.loadFiles(context!!)
                    presenter.source.isLoggedIn = true
                } else if (!shouldShowRequestPermissionRationale(Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
                    Snackbar.make(recycler!!, R.string.storage_permission, Snackbar.LENGTH_LONG)
                        .setAction(R.string.action_settings) { showAppSettings() }
                        .show()
                }
            }
            Constants.RequestCodes.ACCOUNTS_PERMISSIONS -> {
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    if (this is GoogleDriveFragment) {
                        (presenter.source as GoogleDriveSource).authGoogle(this)
                    }
                } else if (!shouldShowRequestPermissionRationale(Manifest.permission.GET_ACCOUNTS)) {
                    Snackbar.make(recycler!!, R.string.contacts_permission, Snackbar.LENGTH_LONG)
                        .setAction(R.string.action_settings) { showAppSettings() }
                        .show()
                }
            }
        }
        (activity as? SourceActivity)?.isCheckingPermissions = false
    }

    override fun showAppSettings() {
        val uri = Uri.fromParts("package", "com.jakebarnby.filemanager", null)
        val intent = Intent().apply {
            action = Settings.ACTION_APPLICATION_DETAILS_SETTINGS
            data = uri
        }
        startActivity(intent)
    }
}